ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+# Ruby SDK - Type hinting, Converters, Codecs and Errors
+
+This proposal goes into details regarding the implementation of data conversion pipeline (including
+type hints, converters and codecs) as well as outlines the hierarchy of exception classes.
+
+## Type hinting
+
+The Ruby typing system is not something we can currently rely on for specifying the types of all the
+inputs and outputs. Instead we will add custom "annotations" (technically these are just method
+calls) to the activity/workflow definitions in order to provide type hinting. Here's the basic
+example:
+
+```ruby
+class SumActivity < Temporal::Activity
+  input Integer, Integer
+  output Integer
+
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+We will support most basic Ruby types (e.g. `Integer`, `Float`, `String`, `Symbol`, `Boolean`,
+`Array`, `Hash`, etc). This will allow the SDK to perform a runtime check of the arguments and
```

Comment:
We should support everything in Ruby that can be serialized/deserialized via standard Ruby JSON (maybe this means we should set `object_class` during `JSON.parse` to this value?). Whatever the Ruby "way" is to JSON deserialize into known class structures we should support.

Reply:
There unfortunately isn't a standard way of doing this.  There's `oj` gem (a popular JSON lib) which has an optional [Object mode](https://github.com/ohler55/oj/blob/develop/pages/Modes.md#object-mode) that I have used in the former SDK, but that obviously won't be compatible with other SDKs

Reply:
> Is there a general "object-from-hash" in Ruby? Sounds like not.

Yep, there isn't.

> "How do I say something is a list of my class w/ the custom load method?"

Potentially we can support the `[MyClass]` notation for the `input`/`output` params. Do you think we should add it straight away?

> So if there are one or two really popular libraries for this in Ruby, we could consider supporting/encouraging as optional dependencies.

Agree, but I don't think there is a consensus on what these options are. I suggest we delay this until we start gathering feedback from the actual users rather than speculate.

Reply:
> Do you think we should add it straight away?

The concern from the Python side was that there are more than arrays, e.g. lists, sets, dictionaries, etc. If this would take care of most people's concerns, sure. Though really we should strongly encourage one single object for every input and output and to make everything else attributes/fields/members/whatever-you-call-them on that object.

Reply:
In that case maybe let's not add it just yet and steer people away from it by not making it easier (they can already implement whatever wrapper they need with the provided interface)